src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
+
+                            session.give((i, batch));
+
+                            // Round robin
+                            i = (i + 1) % peers;
+
+                            if let Some(frontier) = progress {
+                                match frontier.into_option() {
+                                    Some(ts) => {
+                                        cap.downgrade(&ts);
+                                        current_ts = ts;
+                                    }
+                                    None => return SourceStatus::Done,
+                                }
+                            }
+                        }
+                        Some(Err::<_, ExternalError>(e)) => {
+                            panic!("unexpected error from persist {e}")
+                        }
+                        None => return SourceStatus::Done,
+                    }
+
+                    if timer.elapsed() > YIELD_INTERVAL {
+                        activator.activate();
+                        break;
+                    }
+                }
+
+                SourceStatus::Alive
+            }
+        });
+
+    let mut builder = OperatorBuilder::new("partitioned reader".to_string(), scope.clone());
+    let dist = |&(i, _): &(usize, ReaderEnrichedHollowBatch<u64>)| i as u64;
+    let mut input = builder.new_input(&inner, Exchange::new(dist));
+    let (mut output, output_stream) = builder.new_output();
+
+    let mut current_ts = 0;
+
+    builder.build_async(
+        scope.clone(),
+        async_op!(|initial_capabilities, _frontiers| {
+            let persist_client = persist_clients
+                .lock()
+                .await
+                .open(metadata.persist_location.clone())
+                .await
+                .expect("could not open persist client")
+                .clone();
+
+            let read = persist_client
+                .open_reader::<SourceData, (), mz_repr::Timestamp, mz_repr::Diff>(
+                    data_shard.clone(),
+                )
+                .await
+                .expect("could not open persist shard");
+
+            initial_capabilities.clear();
+
+            let mut output_handle = output.activate();
+            while let Some((cap, data)) = input.next() {
+                let mut cap = cap.retain();
+                for (_idx, batch) in data.iter() {
+                    let mut fetcher = read
+                        .hollow_batch_fetcher(batch.clone())
+                        .await
+                        .expect("listen iter cannot fail");
+
+                    for update in fetcher.fetch().await {
+                        match dbg!(update) {
+                            ListenEvent::Progress(upper) => match upper.into_option() {
+                                Some(ts) => {
+                                    current_ts = ts;
+                                    // ??? This seems perfunctory given that
```

Comment:
not 100% sure I understand the question but if it helps, the two operators have different capabilities. you can think of the capabilities of the first stage as something like "here are all the batches one needs to fetch to progress to this point" and the second one like "here are all the actual fetched updates needed to progress to this point"

in fact, I think this suggests we have an opportunity to tighten up the contract between the two stages here:
- the first operator should emit batches and use its output capability to communicate when the full set of batches for some time have been emitted (as it does now)
- this one simply needs to flat_map the batches out by fetching them. concretely: if this operator sees "B1, B2, F1, B3, F2", it simply needs to fetch and emit B1 and B2 before passing on F1 (via cap.downgrade), similarly with B3 and F2. if we decide to in the future, it's safe to begin fetching B3 before we've finished fetching and emitting B1 and B2, we just have to make sure the progress info respects the original order that the batches and frontiers came in
- concretely: this means `HollowBatchFetcher::fetch` (or `ReadHandle::fetch_batch` if we do my comment above) just needs to return updates. it doesn't need to have anything to do with progress or frontiers

Reply:
yes, I think frontier/progress tracking is only a concern of the first-stage operator

Reply:
Ah, this seems to be the comments I posted without reading all the other comments first. I agree with these takes, and my take is purely redundant with these.

Reply:
> concretely: if this operator sees "B1, B2, F1, B3, F2", it simply needs to fetch and emit B1 and B2 before passing on F1 (via cap.downgrade), 

I'm not sure the second operator needs to do `cap.downgrade` at all! It probably just grabs the capability that came with e.g. `B1`, `B2`, and uses it to send the updates for each of the batches, and then just drops the capability. It only needs to hold and downgrade the capability if it plans to emit output updates at times that it will not receive in its input (and I think that is not the case here?).

Reply:
> my take is purely redundant with these.

Still good to see that everyone arrived at the same conclusion! :sweat_smile:

Reply:
> I'm not sure the second operator needs to do cap.downgrade at all! It probably just grabs the capability that came with e.g. B1, B2, and uses it to send the updates for each of the batches, and then just drops the capability. It only needs to hold and downgrade the capability if it plans to emit output updates at times that it will not receive in its input (and I think that is not the case here?).

You are correct of course! I'm so used to only writing source operators at this point that I brain-farted and forgot that incoming data has an attached capability.